# Big O notation

In [1]:
import time

def sumOfN2(n):
   start = time.time()

   theSum = 0
   for i in range(1,n+1):
      theSum = theSum + i

   end = time.time()

   return theSum,end-start

In [4]:
for i in range(5):
       print("Sum is %d required %10.7f seconds"%sumOfN2(10000))

Sum is 50005000 required  0.0010791 seconds
Sum is 50005000 required  0.0010269 seconds
Sum is 50005000 required  0.0010459 seconds
Sum is 50005000 required  0.0011978 seconds
Sum is 50005000 required  0.0011342 seconds


In [8]:
# Self check: write a function that finds the smallest number in a list
# this one is quadratic, two nested loops: O(n)=n^2
import time
from random import randrange
def findmin(alist):
    overallmin = alist[0]
    for i in alist:
        issmallest = True
        for j in alist:
            if i > j:
                issmallest = False
        if issmallest:
            overallmin = i
    return overallmin

In [7]:
print(findmin([5,4,1,8,7]))

1


In [12]:
for listsize in range(1000,10001,1000):
    alist = [randrange(1000000) for x in range(listsize)]
    start = time.time()
    print(findmin(alist))
    end = time.time()
    print('size: %d time: %f' % (listsize, end-start))

2617
size: 1000 time: 0.059557
25
size: 2000 time: 0.217936
32
size: 3000 time: 0.481425
277
size: 4000 time: 0.836147
8
size: 5000 time: 1.368550
28
size: 6000 time: 1.811918
34
size: 7000 time: 2.426557
128
size: 8000 time: 3.150813
106
size: 9000 time: 3.986208
75
size: 10000 time: 4.992156


In [16]:
#this one is linear, O(n)=n
def findmin(alist):
    minsofar = alist[0]
    for i in alist:
        if i< minsofar:
            minsofar = i
    return minsofar

In [18]:
findmin([5,4,3,2,9,0])

0

In [19]:
for listsize in range(1000,10001,1000):
    alist = [randrange(1000000) for x in range(listsize)]
    start = time.time()
    print(findmin(alist))
    end = time.time()
    print('size: %d time: %f' % (listsize, end-start))

1789
size: 1000 time: 0.000217
917
size: 2000 time: 0.000225
333
size: 3000 time: 0.000162
22
size: 4000 time: 0.000212
294
size: 5000 time: 0.000244
152
size: 6000 time: 0.000426
25
size: 7000 time: 0.000327
53
size: 8000 time: 0.000389
18
size: 9000 time: 0.000414
134
size: 10000 time: 0.000416


In [20]:
# Checking off solution to find anagrams
def anagramSolution1(s1,s2):
    alist = list(s2)

    pos1 = 0
    stillOK = True

    while pos1 < len(s1) and stillOK:
        pos2 = 0
        found = False
        while pos2 < len(alist) and not found:
            if s1[pos1] == alist[pos2]:
                found = True
            else:
                pos2 = pos2 + 1

        if found:
            alist[pos2] = None
        else:
            stillOK = False

        pos1 = pos1 + 1

    return stillOK

print(anagramSolution1('abcd','dcba'))

True


In [25]:
print(anagramSolution1('marlene','enelram'))

True


In [40]:
# The sorting solution to find anagrams:
def anagramSolution2(s1,s2):
    alist1 = list(s1)
    alist2 = list(s2)

    alist1.sort()
    alist2.sort()

    pos = 0
    matches = True

    while pos < len(s1) and matches:
        if alist1[pos]==alist2[pos]:
            pos = pos + 1
        else:
            matches = False

    return matches

print(anagramSolution2('abcde','edcba'))

True


In [51]:
# The counting solution to find anagrams
def anagramSolution4(s1,s2):
    c1 = [0]*26
    c2 = [0]*26
# ord() Given a string representing one Unicode character, return an integer representing the Unicode 
#code point of that character. For example, ord('a') returns the integer 97 and ord('€') (Euro sign) 
#returns 8364. This is the inverse of chr().
    for i in range(len(s1)):
        pos = ord(s1[i])-ord('a')
        c1[pos] = c1[pos] + 1

    for i in range(len(s2)):
        pos = ord(s2[i])-ord('a')
        c2[pos] = c2[pos] + 1
    if c1 == c2:
        return True
    else:
        return False
#Don't know why he needs this long thing, in the end, if they're anagrams ci should be equal to c2, no?
#     j = 0
#     stillOK = True
#     while j<26 and stillOK:
#         if c1[j]==c2[j]:
#             j = j + 1
#         else:
#             stillOK = False

#     return stillOK


In [55]:
print(anagramSolution4('mirlene','enelram'))

False


In [60]:
chr(97)

'a'

# Lists

Four different ways to grow a list of n numbers starting with 0. 

In [61]:
def test1():
    l = []
    for i in range(1000):
        l = l + [i]

def test2():
    l = []
    for i in range(1000):
        l.append(i)

def test3():
    l = [i for i in range(1000)]

def test4():
    l = list(range(1000))


Now let's time them:

In [69]:
import timeit as ti
t1 = ti.Timer("test1()", "from __main__ import test1")
print("concat ",t1.timeit(number=1000), "milliseconds")
t2 = ti.Timer("test2()", "from __main__ import test2")
print("append ",t2.timeit(number=1000), "milliseconds")
t3 = ti.Timer("test3()", "from __main__ import test3")
print("comprehension ",t3.timeit(number=1000), "milliseconds")
t4 = ti.Timer("test4()", "from __main__ import test4")
print("list range ",t4.timeit(number=1000), "milliseconds")

concat  1.8148007929848973 milliseconds
append  0.11429262300953269 milliseconds
comprehension  0.0495977149985265 milliseconds
list range  0.02180514100473374 milliseconds


**NOTE:** The statement from ``__main__ import test1`` imports the function test1 from the `__main__` namespace into the namespace that `timeit` sets up for the timing experiment. The timeit module does this because it wants to run the timing tests in an environment that is uncluttered by any stray variables you may have created, that may interfere with your function’s performance in some unforeseen way.

In [75]:
def hola():
    pass
te = ti.Timer('hola()','from __main__ import hola')
print("empty function", te.timeit(number=1000),'milliseconds')

empty function 0.00012238198542036116 milliseconds


**Note2:** It's not exactly accurate to say that concat takes 1.81 milliseconds, for example, bc python takes some time to call the function. The line above illustrates how much time it takes to call an empty funciton. 

In [77]:
popzero = timeit.Timer("x.pop(0)",
                       "from __main__ import x")
popend = timeit.Timer("x.pop()",
                      "from __main__ import x")

x = list(range(2000000))

In [78]:
popzero.timeit(number=1000)

2.1711451269802637

In [80]:
popend.timeit(number=1000)

0.0001390129909850657

So, popend is indeed a lot faster than popzero, but the claim here is that ``popzero`` is $O(n)$, while ``popend`` is $O(1)$, so that has to do with the rates of change as n grows (one should be constant, the other increasing), let's check:

In [82]:
popzero = ti.Timer("x.pop(0)",
                "from __main__ import x")
popend = ti.Timer("x.pop()",
               "from __main__ import x")
print("pop(0)   pop()")
for i in range(1000000,100000001,1000000):
    x = list(range(i))
    pt = popend.timeit(number=1000)
    x = list(range(i))
    pz = popzero.timeit(number=1000)
    print("%15.5f, %15.5f" %(pz,pt))


pop(0)   pop()
        1.13644,         0.00014
        1.99578,         0.00014
        2.91566,         0.00034
        3.98653,         0.00013
        4.78120,         0.00036
        5.89120,         0.00013
        6.79368,         0.00021
        8.15829,         0.00035
        9.12003,         0.00035
        9.96094,         0.00039
       10.76452,         0.00037
       12.43674,         0.00013
       13.77721,         0.00014
       16.36170,         0.00015
       20.18388,         0.00016
       20.64745,         0.00014
       20.65110,         0.00029
       20.54851,         0.00016


KeyboardInterrupt: 

# Dictionaries

In [87]:
import random
import timeit
i = 50
x = {j:None for j in range(i)}

In [89]:
for i in range(10000,1000001,200000):
    t = timeit.Timer("random.randrange(%d) in x"%i,
                     "from __main__ import random,x")
    x = list(range(i))
    lst_time = t.timeit(number=1000)
    x = {j:None for j in range(i)}
    d_time = t.timeit(number=1000)
    print("%d,%10.3f,%10.3f" % (i, lst_time, d_time))


10000,     0.120,     0.002
210000,     2.272,     0.002
410000,     4.338,     0.002
610000,     6.173,     0.002
810000,     8.498,     0.002
